File: typescript/base/src/main/vsdk/toolkit/common/commandLineOptions/DefaultParser.ts
import { TypedOption } from "./TypedOption";

export class DefaultParser {
  private constructor() {
  }

  public static parse<T>(input: string | null, out: TypedOption.MutableValue<T> | null): boolean {
    if (input === null || out === null) {
      return false;
    }

    const currentValue = out.value as unknown;

    if (typeof currentValue === "number") {
      if (Number.isInteger(currentValue)) {
        const converted = new TypedOption.MutableValue<number>(currentValue);
        if (DefaultParser.parseInt(input, converted)) {
          out.value = converted.value as T;
          return true;
        }
      }

      const converted = new TypedOption.MutableValue<number>(currentValue);
      if (!DefaultParser.parseFloat(input, converted)) {
        return false;
      }
      out.value = converted.value as T;
      return true;
    }

    if (typeof currentValue === "bigint") {
      const converted = new TypedOption.MutableValue<bigint>(currentValue);
      if (!DefaultParser.parseLong(input, converted)) {
        return false;
      }
      out.value = converted.value as T;
      return true;
    }

    if (typeof currentValue === "boolean") {
      const boolValue = DefaultParser.parseBoolean(input);
      if (boolValue === null) {
        return false;
      }
      out.value = boolValue as T;
      return true;
    }

    out.value = input as T;
    return true;
  }

  private static parseInt(input: string, out: TypedOption.MutableValue<number> | null): boolean {
    if (input === null || out === null) {
      return false;
    }

    const parsedValue = DefaultParser.parseInteger(input);
    if (parsedValue === null) {
      return false;
    }
    if (parsedValue < -2147483648 || parsedValue > 2147483647) {
      return false;
    }

    out.value = parsedValue;
    return true;
  }

  private static parseLong(input: string, out: TypedOption.MutableValue<bigint> | null): boolean {
    if (input === null || out === null) {
      return false;
    }
    if (!/^[+-]?\d+$/.test(input)) {
      return false;
    }

    try {
      out.value = BigInt(input);
      return true;
    }
    catch (_e) {
      return false;
    }
  }

  private static parseFloat(input: string, out: TypedOption.MutableValue<number> | null): boolean {
    if (input === null || out === null) {
      return false;
    }

    const parsedValue = DefaultParser.parseNumber(input);
    if (parsedValue === null) {
      return false;
    }

    out.value = parsedValue;
    return true;
  }

  private static parseInteger(input: string): number | null {
    if (!/^[+-]?\d+$/.test(input)) {
      return null;
    }

    const parsedValue = Number.parseInt(input, 10);
    if (!Number.isFinite(parsedValue)) {
      return null;
    }
    return parsedValue;
  }

  private static parseNumber(input: string): number | null {
    if (input.trim().length === 0) {
      return null;
    }
    const parsedValue = Number(input);
    if (!Number.isFinite(parsedValue)) {
      return null;
    }
    return parsedValue;
  }

  private static parseBoolean(input: string): boolean | null {
    if (
      DefaultParser.equalsIgnoreCase(input, "true")
      || DefaultParser.equalsIgnoreCase(input, "yes")
      || input === "1"
    ) {
      return true;
    }

    if (
      DefaultParser.equalsIgnoreCase(input, "false")
      || DefaultParser.equalsIgnoreCase(input, "no")
      || input === "0"
    ) {
      return false;
    }

    return null;
  }

  private static toLowerAscii(c: string): string {
    if (c >= "A" && c <= "Z") {
      return String.fromCharCode(c.charCodeAt(0) - "A".charCodeAt(0) + "a".charCodeAt(0));
    }
    return c;
  }

  private static equalsIgnoreCase(a: string | null, b: string | null): boolean {
    if (a === null || b === null) {
      return false;
    }

    if (a.length !== b.length) {
      return false;
    }

    for (let i = 0; i < a.length; i++) {
      if (DefaultParser.toLowerAscii(a.charAt(i)) !== DefaultParser.toLowerAscii(b.charAt(i))) {
        return false;
      }
    }
    return true;
  }
}

File: java/base/src/main/java/vsdk/toolkit/common/commandLineOptions/DefaultParser.java
package vsdk.toolkit.common.commandLineOptions;

public class DefaultParser<T> {
    @SuppressWarnings("unchecked")
    public static <T> boolean parse(String input, TypedOption.MutableValue<T> out) {
        if (input == null || out == null) {
            return false;
        }

        Object currentValue = out.value;

        if (currentValue instanceof Integer) {
            TypedOption.MutableValue<Integer> converted = new TypedOption.MutableValue<>((Integer)currentValue);
            if (!parseInt(input, converted)) {
                return false;
            }
            out.value = (T)converted.value;
            return true;
        }
        if (currentValue instanceof Long) {
            TypedOption.MutableValue<Long> converted = new TypedOption.MutableValue<>((Long)currentValue);
            if (!parseLong(input, converted)) {
                return false;
            }
            out.value = (T)converted.value;
            return true;
        }
        if (currentValue instanceof Float) {
            TypedOption.MutableValue<Float> converted = new TypedOption.MutableValue<>((Float)currentValue);
            if (!parseFloat(input, converted)) {
                return false;
            }
            out.value = (T)converted.value;
            return true;
        }
        if (currentValue instanceof Boolean) {
            TypedOption.MutableValue<Boolean> converted = new TypedOption.MutableValue<>((Boolean)currentValue);
            if (!parseBoolean(input, converted)) {
                return false;
            }
            out.value = (T)converted.value;
            return true;
        }

        out.value = (T)input;
        return true;
    }

    private static boolean parseInt(String input, TypedOption.MutableValue<Integer> out) {
        if (input == null || out == null) {
            return false;
        }

        try {
            long parsedValue = Long.parseLong(input, 10);
            if (parsedValue < Integer.MIN_VALUE || parsedValue > Integer.MAX_VALUE) {
                return false;
            }
            out.value = (int)parsedValue;
            return true;
        }
        catch (NumberFormatException e) {
            return false;
        }
    }

    private static boolean parseLong(String input, TypedOption.MutableValue<Long> out) {
        if (input == null || out == null) {
            return false;
        }

        try {
            out.value = Long.parseLong(input, 10);
            return true;
        }
        catch (NumberFormatException e) {
            return false;
        }
    }

    private static boolean parseFloat(String input, TypedOption.MutableValue<Float> out) {
        if (input == null || out == null) {
            return false;
        }

        try {
            out.value = Float.parseFloat(input);
            return true;
        }
        catch (NumberFormatException e) {
            return false;
        }
    }

    private static boolean parseBoolean(String input, TypedOption.MutableValue<Boolean> out) {
        if (input == null || out == null) {
            return false;
        }

        if (equalsIgnoreCase(input, "true")
            || equalsIgnoreCase(input, "yes")
            || "1".equals(input)) {
            out.value = true;
            return true;
        }

        if (equalsIgnoreCase(input, "false")
            || equalsIgnoreCase(input, "no")
            || "0".equals(input)) {
            out.value = false;
            return true;
        }

        return false;
    }

    private static char toLowerAscii(char c) {
        if (c >= 'A' && c <= 'Z') {
            return (char)(c - 'A' + 'a');
        }
        return c;
    }

    private static boolean equalsIgnoreCase(String a, String b) {
        if (a == null || b == null) {
            return false;
        }

        if (a.length() != b.length()) {
            return false;
        }

        for (int i = 0; i < a.length(); i++) {
            if (toLowerAscii(a.charAt(i)) != toLowerAscii(b.charAt(i))) {
                return false;
            }
        }
        return true;
    }
}
